File: Java_DBFundamentals-Jan18/DBAdvanced_HibernateSpring/d_Java_Oop_Overview/src/main/java/e_IntersectionOfCircles/Circle.java
package e_IntersectionOfCircles;

import java.math.BigDecimal;

public class Circle {

    private Point center;
    private BigDecimal radius;

    public Circle(String x, String y, String r){
        this.setCenter(x, y);
        this.setRadius(r);
    }

    public Circle(String[] xyr) {
        this.setCenter(xyr[0], xyr[1]);
        this.setRadius(xyr[2]);
    }

    public void setCenter(String x, String y) {
        this.center = new Point(x, y);
    }

    public void setRadius(String r) {
        this.radius = new BigDecimal(r);
    }

    public Point getCenter() {
        return this.center;
    }

    public BigDecimal getRadius() {
        return this.radius;
    }

    public double getRadiusAsDouble(){
        return this.radius.doubleValue();
    }

    public double getArea(){
        /* pi*r^2 */
        return Math.PI * this.radius.pow(2).doubleValue();
    }

    public double getCircumference(){
        /* 2 * pi * r          */
        return 2.0 * Math.PI * this.getRadiusAsDouble();
    }

    public boolean intersectWithCircle(Circle otherCircle){
        Double centresDistance = this.center.getDistanceToPoint(otherCircle.getCenter());
        Double sumOfRadiuses = this.radius.add(otherCircle.radius).doubleValue();

        return sumOfRadiuses >= centresDistance;
    }
}

File: JSCore/JSAdvanced/l_JSClasses_lab/d_Circle/circle.js
class Circle{
    constructor (radius){ this.radius = radius; }
    get diameter() { return this.radius * 2; }
    set diameter(diam) { this.radius = diam / 2; }
    get area() { return Math.PI * this.radius ** 2 }
}

let c = new Circle(2);
console.log(`Radius: ${c.radius}`); // 2
console.log(`Diameter: ${c.diameter}`); // 4
console.log(`Area: ${c.area}`); // 12.566370614359172

c.diameter = 1.6;
console.log(`Radius: ${c.radius}`); // 0.8
console.log(`Diameter: ${c.diameter}`); // 1.6
console.log(`Area: ${c.area}`); // 2.0106192982974678